Feature/degree template

sarthak22gaur opened the pull request:
## Overview

Feature added for creating plan from template.

## What Changed

`newhome/Home.tsx` modified to intercept new plan handler and forward to template Modal.
Added `api/template.ts` to return an object of courses in place of an array for better performance.
`components/template/Modal.tsx` handles the plan creation

## TODO
1. Add all templates to `degree-template.json`. Only Psychology and Computer Science added as of now. (A lot of manual entry which I was lazy to do right now)
2. Improve error handling in case course is not found, at the moment errors are handled by discarding the particular entry.
3. Fix hard-coded credits for elective courses

sarthak22gaur committed:
feat(template component): Templates for degree plan

Templates and template components for creating new degree plans

BREAKING CHANGE: New plan flow is changed

sarthak22gaur committed:
feat(template): Create plan from Template

On creating a new plan, provide options to chose from predefined templates.

ZzRanger commented on components/template/Modal.tsx:
```diff
@@ -0,0 +1,124 @@
+        } else {
+          try {
+            console.log(c);
+            const { id, name: title, description, hours } = courses[c];
```
Can we not include course if user already has it in creditSlice?

sarthak22gaur replied:
This is a bigger issue that will require some extra stuff as of now. 
1. For including already completed courses, we need to remove identical/same courses from the template, which will result in pre-req requirements going haywire if they are not in the right semesters initially.
2. Electives or for the courses that provide options like '060', '050' etc, i don't think there is a way to identify those from the credits list (I might be wrong on this).

**Implementing this properly so that it suffices all criteria needs a little more work in terms of how courses are structured and labelled based on departments and electives and so on.**

ZzRanger replied:
Makes sense; I think for the time being I'm okay with a scuffed design and we can develop a better solution later on

**For now**
1. Do you mind clarifying your point? Might just be me but I don't fully understand what you're saying. I assume if we add the courses from creditSlice into the plan first (putting the courses into the prior semesters), then add the template plan into the current semesters (i.e. start w/ Spring 2023) and remove duplicate courses in template plan, it should work?

2. I'm perfectly fine with not checking for option courses like '060', '050' and letting the user manually validate. 

Alternatively, I think we can easily check for the numerical options. A list of all the courses for each of the numerical options can be found [here](https://github.com/ez314/Degree-Validator/blob/main/requirements/computer_science_ug.req). If we want to check if the user has satisfied that credit we can check if the user has already taken a course in the list. For other types like "Elective" or "CS Elective", I think we can effectively ignore this for now. Let me know what your thoughts are on this.

**Future**
We are using this [API](https://github.com/ez314/Degree-Validator) to validate degree requirements. I think the ideal implementation for generating template degree plans could use this API to know which courses the user still needs to take (i.e. we can send a POST request with all the courses in creditsSlice to ascertain what requirements the user still needs to satisfy, then using that information we can generate a personalized template degree plan that suits their needs).

Let me know what your thoughts are on this.

sarthak22gaur replied:
There are mainly two issues,
1. Suppose I have a course in the creditSlice and that is added to the degree plan first, but that course is not there in the template, in this case, I'll have more courses than the required/allowed courses because I'll have all the template courses plus the additional creditSlice courses. 
2. This is not that big of an issue but if the creditSlice already has any of the elective/"050" or similar courses, then still the template will have those empty banners for those courses.

- **Possible implementation: Just dump everything and put a banner that there could be duplication of courses  and let the user figure it out at least for the beta release and then add all the validation and everything incrementally for the alpha release**

### Should all the courses from the creditSlice go into the first semester or separate them into different semesters somehow? 

#### From what I have briefly seen from the API, I think the data from that will help resolve a lot of the issues, we can integrate that after the beta release.

ZzRanger replied:
Those are great concerns you brought up! I agree with the possible implementation of just dumping courses for now. 1) isn't that big of a deal and 2) we can live with for now.

To answer your question, I think all the courses from creditSlice should be separated into different semesters if possible. One way we can do that off the top of my head is to create an object that maps semester name to Semester objects that we can add courses in. Then, we can grab the semester objects and add that to the beginning of the degree plan. Let me know what you think about this approach.

sarthak22gaur replied:
I don't quite understand the solution you proposed, so for now I've distributed the courses from `creditSlice` equally among all semesters.

ZzRanger replied:
My apologies I might not have communicated it super well. [This](https://github.com/UTDNebula/planner/blob/chore/template-show/components/template/Modal.tsx#L117-L119) is what I was thinking from above.

ZzRanger replied:
![Oct-28-2022 18-46-27](https://user-images.githubusercontent.com/47403443/198751257-b77bc4e0-91c5-42d5-8e3b-01b831f4486b.gif)

ZzRanger reviewed (commented):
Overall looks good, great work! I'll see if I can find people to do manual entry for the sample degree plans. I like the progress that you are making on this feature!

sarthak22gaur committed:
refactor(template/modal.tsx): Removed emptyCourse, moved away from map

sarthak22gaur commented on components/newhome/Home.tsx:
```diff
@@ -50,24 +36,28 @@ export default function PlansPage(): JSX.Element {
+        </section>
+      </main>
+
+      {openTemplateModal && <TemplateModal setModal={setModal} />}
```
It only sets the Modal, I'm just not in the habit of passing the state callback so I created a function.
Removed this in the latest commit.

sarthak22gaur commented on components/template/Modal.tsx:
```diff
@@ -0,0 +1,124 @@
+  options: string;
+};
+
+const emptyCourse: Course = {
```
Moved away from this completely, this was a necessity while using the `Array.map` iteration, removed that as well and moved to a simple `for` loop which also prevents unnecessary array duplication.

sarthak22gaur commented on components/template/Modal.tsx:
```diff
@@ -0,0 +1,124 @@
+  const router = useRouter();
+  const dispatch = useDispatch();
+
+  const handleTemplateCreation = async (val: string) => {
```
Done

sarthak22gaur commented on components/template/Modal.tsx:
```diff
@@ -0,0 +1,124 @@
+    let season = 'Fall';
+    dummyTemplate[val].map((sem, index) => {
+      const semCourses: Course[] = [];
+      const semTitle = `${season} ${year + Math.floor(index / 2)}`;
```
I'll look into this, need to do a few reactors to make the arguments typesafe.

sarthak22gaur replied:
cannot use `generateSemesters` in its current form, so I'm keeping the logic in the Modal only for now. We can move this to a server component later as it is the most ideal place for it.

sarthak22gaur commented on components/template/Modal.tsx:
```diff
@@ -0,0 +1,124 @@
+              description,
+              creditHours: +hours,
+            };
+          } catch (e) {
```
Don't need it anymore, it was needed when using the `map` iterator because it creates a null object otherwise.
Also don't need the `filter` now.
Fixed in the latest commit

sarthak22gaur committed:
feat(template/modal.tsx): Implemented course addition from creditSlice

sarthak22gaur commented on components/template/Modal.tsx:
```diff
@@ -91,12 +136,22 @@ export default function TemplateModal({ setOpenTemplateModal }: TemplateModalPro
+        className="bg-white relative max-w-4xl m-8 max-h-[90vh] w-full overflow-y-scroll flex-col gap-2 rounded-lg items-center justify-center p-4"
       >
-        <div className="text-center text-2xl px-2 py-4">Select a Template for the degree plan</div>
+        <div className="flex justify-center items-center">
+          <div className="p-2 text-2xl">Start fresh with an</div>
+          <button
+            className="p-2 rounded hover:bg-[#3E61ED] hover:text-white transition-colors border-2 border-[#3e61ed]"
+            onClick={() => handleTemplateCreation('empty')}
+          >
+            Empty Plan
+          </button>
+        </div>
+        <div className="text-center p-2 text-4xl font-bold">OR</div>
+        <div className="text-center text-2xl p-2">Start with one of the templates</div>
```
Added option for starting from a blank template if someone prefers that.

ZzRanger commented:
<img width="1440" alt="Screen Shot 2022-10-28 at 6 38 43 PM" src="https://user-images.githubusercontent.com/47403443/198750805-f728f38f-f1f2-49e3-9237-ea4f5ce00c0c.png">

I noticed a bug: Spring 2022 should not come after Fall 2022 and so on and so forth.

sarthak22gaur committed:
feat(template/modal.tsx): Added user credits to degree template

sarthak22gaur committed:
chore(newhome/home.tsx): Changed plancard styling to prevent overflow

changed the PlanCard to be more responsive and prevent overflow.

sarthak22gaur committed:
feat(generic warningmodal): A generic expanding warning message modal

An optional warning modal added on the degree plan to notify of possible
discrepencies in the template

ZzRanger committed:
Merge branch 'develop' into feature/degree-template

ZzRanger reviewed (approved):
LGTM!
